### ISSUE TITLE
[Proposal] Allow methods named Invoke to be invoked without the method name.

### ISSUE THREAD

**Issue by erik-kallen** [dotnet/roslyn#1611]
It would be neat that, given

```
class C {
    public void Invoke(int a) {
    }
}
...
var c = new C();
```

you could write ``` c(42)``` instead of ```c.Invoke(42)```. This would be a generalization of the pattern that is currently only valid for delegates do be allowed for all objects.

**Comment by paulomorgado** [comment]
Is this proposal only for C# or also for Visual Basic?

**Comment by AdamSpeight2008** [comment]
How does this affect indexers?

**Comment by erik-kallen** [comment]
@paulomorgado Since I only use C# I only propose it for that language. It might, however, be useful for VB too but I don't know if it might cause ambiguities or other problems.

@AdamSpeight2008 I don't see how indexers would need to be affected at all. The only way I see the features interact is if you use ```[IndexerName]``` to create an indexer named ```Invoke```, in which case it is not a method so the indexer takes precedence.

**Comment by svick** [comment]
@erik-kallen My guess is that @AdamSpeight2008 is talking about VB, where `()` are used for both indexers and delegate invocation.

**Comment by svick** [comment]
What exactly is the point of this feature? C++ and Java both have a similar feature, but I think that's mostly because to make lambdas work with existing patterns (functors in C++, functional interfaces in Java), not the other way around.

**Comment by AdamSpeight2008** [comment]
Also `()` is used for method invocation.

**Comment by KalitaAlexey** [comment]
I think that's have no any benefits. If it would be implemented, then people will try use this classes as parameters of Action type, but that's wrong. It confuses.

**Comment by gafter** [comment]
I don't see why this would be neat. ELI5.

It seems like indexers gets you the syntactic convenience without making the code confusing.

**Comment by KalitaAlexey** [comment]
@gafter One difference between Invocable class and Action is ability to save data between calls.
I am asking again: "Can people use invocable class as Action parameters?"

**Comment by erik-kallen** [comment]
Perhaps it is unnecessary, my thinkings were along the lines that we already have this pattern in the language (but only for types that inherit ```System.Delegate```), so why not remove the inheritance restriction? But as @gafter says, since the name lookup rules are different for invocations compared to non-invocations, the change probably isn't as simple as I thought.

**Comment by gafter** [comment]
@KalitaAlexey A delegate of type Action certainly can save data between calls.

We cannot answer your question "Can people use invocable class as Action parameters?" without knowing what you mean by "invocable class".
